File: python/Komputer.py
class Komputer:
    def __init__(self, nama="", kategori="", motherboard=None, cpu=None,
                 ram_list=None, storage_list=None, gpu_list=None,
                 power_supply=None, casing=None, cooling_list=None):
        self.nama = nama
        self.kategori = kategori
        self.motherboard = motherboard
        self.cpu = cpu
        self.ram_list = ram_list if ram_list is not None else []
        self.storage_list = storage_list if storage_list is not None else []
        self.gpu_list = gpu_list if gpu_list is not None else []
        self.power_supply = power_supply
        self.casing = casing
        self.cooling_list = cooling_list if cooling_list is not None else []
        self.total_harga = 0.0
        self.total_power_consumption = 0
        self.calculate_total_harga()
        self.calculate_total_power_consumption()
    
    def calculate_total_harga(self):
        self.total_harga = 0.0
        if self.motherboard:
            self.total_harga += self.motherboard.get_harga()
        if self.cpu:
            self.total_harga += self.cpu.get_harga()
        self.total_harga += sum(ram.get_harga() for ram in self.ram_list)
        self.total_harga += sum(storage.get_harga() for storage in self.storage_list)
        self.total_harga += sum(gpu.get_harga() for gpu in self.gpu_list)
        if self.power_supply:
            self.total_harga += self.power_supply.get_harga()
        if self.casing:
            self.total_harga += self.casing.get_harga()
        self.total_harga += sum(cooling.get_harga() for cooling in self.cooling_list)
    
    def calculate_total_power_consumption(self):
        self.total_power_consumption = 0
        if self.motherboard:
            self.total_power_consumption += self.motherboard.get_power_consumption()
        if self.cpu:
            self.total_power_consumption += self.cpu.get_power_consumption()
        self.total_power_consumption += sum(ram.get_power_consumption() for ram in self.ram_list)
        self.total_power_consumption += sum(storage.get_power_consumption() for storage in self.storage_list)
        self.total_power_consumption += sum(gpu.get_power_consumption() for gpu in self.gpu_list)
        self.total_power_consumption += sum(cooling.get_power_consumption() for cooling in self.cooling_list)
    
    def is_power_sufficient(self):
        return self.power_supply and self.power_supply.get_wattage() >= self.total_power_consumption
    
    def display_info(self):
        print("=" * 50)
        print("COMPUTER INFORMATION")
        print("=" * 50)
        print(f"Name: {self.nama}")
        print(f"Category: {self.kategori}")
        print(f"Total Price: Rp{self.total_harga:,.0f}")
        print(f"Total Power Consumption: {self.total_power_consumption}W")
        print(f"Power Supply Sufficient: {'Yes' if self.is_power_sufficient() else 'No'}")
        print()
        
        if self.motherboard:
            print("MOTHERBOARD:")
            self.motherboard.display_info()
            print()
        
        if self.cpu:
            print("CPU:")
            self.cpu.display_info()
            print()
        
        print(f"RAM MODULES ({len(self.ram_list)}):")
        for i, ram in enumerate(self.ram_list, 1):
            print(f"RAM {i}:")
            ram.display_info()
            print()
        
        print(f"STORAGE DEVICES ({len(self.storage_list)}):")
        for i, storage in enumerate(self.storage_list, 1):
            print(f"Storage {i} ({storage.get_drive_type()}):")
            storage.display_info()
            print()
        
        print(f"GRAPHICS CARDS ({len(self.gpu_list)}):")
        for i, gpu in enumerate(self.gpu_list, 1):
            print(f"GPU {i}:")
            gpu.display_info()
            print()
        
        if self.power_supply:
            print("POWER SUPPLY:")
            self.power_supply.display_info()
            print()
        
        if self.casing:
            print("CASE:")
            self.casing.display_info()
            print()
        
        print(f"COOLING SYSTEMS ({len(self.cooling_list)}):")
        for i, cooling in enumerate(self.cooling_list, 1):
            print(f"Cooling {i}:")
            cooling.display_info()
            print()
        
        print("=" * 50)

File: cpp/Komputer.cpp
#pragma once
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include "Cpu.cpp"
#include "Ram.cpp"
#include "Harddrive.cpp"
#include "SSD.cpp"
#include "GPU.cpp"
#include "Motherboard.cpp"
#include "PowerSupply.cpp"
#include "Case.cpp"
#include "Cooling.cpp"

class Komputer {
private:
    string nama;
    string kategori;
    Motherboard motherboard;
    Cpu cpu;
    vector<Ram> ramList;
    vector<StorageDrive*> storageList;
    vector<GPU> gpuList;
    PowerSupply powerSupply;
    Case casing;
    vector<Cooling> coolingList;
    double totalHarga;
    int totalPowerConsumption;

    // Private method to calculate total price of all components
    void calculateTotalHarga() {
        totalHarga = 0.0;
        
        // Add price of all components
        totalHarga += motherboard.getHarga();
        totalHarga += cpu.getHarga();
        
        // Add RAM prices
        for (const Ram& ram : ramList) {
            totalHarga += ram.getHarga();
        }
        
        // Add storage prices
        for (StorageDrive* storage : storageList) {
            totalHarga += storage->getHarga();
        }
        
        // Add GPU prices
        for (const GPU& gpu : gpuList) {
            totalHarga += gpu.getHarga();
        }
        
        totalHarga += powerSupply.getHarga();
        totalHarga += casing.getHarga();
        
        // Add cooling prices
        for (const Cooling& cooling : coolingList) {
            totalHarga += cooling.getHarga();
        }
    }
    
    // Private method to calculate total power consumption
    void calculateTotalPowerConsumption() {
        totalPowerConsumption = 0;
        
        // Add power consumption of all components
        totalPowerConsumption += motherboard.getPowerConsumption();
        totalPowerConsumption += cpu.getPowerConsumption();
        
        // Add RAM power consumption
        for (const Ram& ram : ramList) {
            totalPowerConsumption += ram.getPowerConsumption();
        }
        
        // Add storage power consumption
        for (StorageDrive* storage : storageList) {
            totalPowerConsumption += storage->getPowerConsumption();
        }
        
        // Add GPU power consumption
        for (const GPU& gpu : gpuList) {
            totalPowerConsumption += gpu.getPowerConsumption();
        }
        
        // Add cooling system power consumption
        for (const Cooling& cooling : coolingList) {
            totalPowerConsumption += cooling.getPowerConsumption();
        }
    }

public:
    Komputer() : nama(""), kategori(""), totalHarga(0.0), totalPowerConsumption(0) {}
    
    Komputer(string nama, string kategori, Motherboard motherboard, Cpu cpu, 
             vector<Ram> ramList, vector<StorageDrive*> storageList, vector<GPU> gpuList,
             PowerSupply powerSupply, Case casing, vector<Cooling> coolingList) {
        this->nama = nama;
        this->kategori = kategori;
        this->motherboard = motherboard;
        this->cpu = cpu;
        this->ramList = ramList;
        this->storageList = storageList;
        this->gpuList = gpuList;
        this->powerSupply = powerSupply;
        this->casing = casing;
        this->coolingList = coolingList;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void setNama(string nama) {
        this->nama = nama;
    }
    
    void setKategori(string kategori) {
        this->kategori = kategori;
    }
    
    void setMotherboard(Motherboard motherboard) {
        this->motherboard = motherboard;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void setCpu(Cpu cpu) {
        this->cpu = cpu;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void setRamList(vector<Ram> ramList) {
        this->ramList = ramList;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void setStorageList(vector<StorageDrive*> storageList) {
        this->storageList = storageList;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void setGpuList(vector<GPU> gpuList) {
        this->gpuList = gpuList;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void setPowerSupply(PowerSupply powerSupply) {
        this->powerSupply = powerSupply;
        calculateTotalHarga();
    }
    
    void setCasing(Case casing) {
        this->casing = casing;
        calculateTotalHarga();
    }
    
    void setCoolingList(vector<Cooling> coolingList) {
        this->coolingList = coolingList;
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void addRam(Ram ram) {
        this->ramList.push_back(ram);
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void addStorage(StorageDrive* storage) {
        this->storageList.push_back(storage);
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void addGpu(GPU gpu) {
        this->gpuList.push_back(gpu);
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    void addCooling(Cooling cooling) {
        this->coolingList.push_back(cooling);
        calculateTotalHarga();
        calculateTotalPowerConsumption();
    }
    
    string getNama() const {
        return nama;
    }
    
    string getKategori() const {
        return kategori;
    }
    
    Motherboard getMotherboard() const {
        return motherboard;
    }
    
    Cpu getCpu() const {
        return cpu;
    }
    
    vector<Ram> getRamList() const {
        return ramList;
    }
    
    vector<StorageDrive*> getStorageList() const {
        return storageList;
    }
    
    vector<GPU> getGpuList() const {
        return gpuList;
    }
    
    PowerSupply getPowerSupply() const {
        return powerSupply;
    }
    
    Case getCasing() const {
        return casing;
    }
    
    vector<Cooling> getCoolingList() const {
        return coolingList;
    }
    
    double getTotalHarga() const {
        return totalHarga;
    }
    
    int getTotalPowerConsumption() const {
        return totalPowerConsumption;
    }
    
    bool isPowerSufficient() const {
        return powerSupply.getWattage() >= totalPowerConsumption;
    }
    
    // Method to display all computer information
    void displayInfo() const {
        cout << "====================================================" << endl;
        cout << "COMPUTER INFORMATION" << endl;
        cout << "====================================================" << endl;
        cout << "Name: " << nama << endl;
        cout << "Category: " << kategori << endl;
        cout << fixed << setprecision(0) << "Total Price: Rp" << totalHarga << endl;
        cout << "Total Power Consumption: " << totalPowerConsumption << "W" << endl;
        cout << "Power Supply Sufficient: " << (isPowerSufficient() ? "Yes" : "No") << endl;
        cout << endl;
        
        cout << "MOTHERBOARD:" << endl;
        motherboard.displayInfo();
        cout << endl;
        
        cout << "CPU:" << endl;
        cpu.displayInfo();
        cout << endl;
        
        cout << "RAM MODULES (" << ramList.size() << "):" << endl;
        for (size_t i = 0; i < ramList.size(); i++) {
            cout << "RAM " << (i + 1) << ":" << endl;
            ramList[i].displayInfo();
            cout << endl;
        }
        
        cout << "STORAGE DEVICES (" << storageList.size() << "):" << endl;
        for (size_t i = 0; i < storageList.size(); i++) {
            cout << "Storage " << (i + 1) << " (" << storageList[i]->getDriveType() << "):" << endl;
            storageList[i]->displayInfo();
            cout << endl;
        }
        
        cout << "GRAPHICS CARDS (" << gpuList.size() << "):" << endl;
        for (size_t i = 0; i < gpuList.size(); i++) {
            cout << "GPU " << (i + 1) << ":" << endl;
            gpuList[i].displayInfo();
            cout << endl;
        }
        
        cout << "POWER SUPPLY:" << endl;
        powerSupply.displayInfo();
        cout << endl;
        
        cout << "CASE:" << endl;
        casing.displayInfo();
        cout << endl;
        
        cout << "COOLING SYSTEMS (" << coolingList.size() << "):" << endl;
        for (size_t i = 0; i < coolingList.size(); i++) {
            cout << "Cooling " << (i + 1) << ":" << endl;
            coolingList[i].displayInfo();
            cout << endl;
        }
        
        cout << "====================================================" << endl;
    }

};